# Название проекта: Исследование надежности заемщиков

**Описание исследования.**

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

**Цель исследования.**

Необходимо исследовать данные и ответить на вопросы:

- Есть ли зависимость между количеством детей и возвратом кредита в срок?
- Есть ли зависимость между семейным положением и возвратом кредита в срок?
- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
- Как разные цели кредита влияют на его возврат в срок?
- Приведите возможные причины появления пропусков в исходных данных.
- Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.


**Задачи исследования.**

- выгрузить датасеты и изучить по ними общую информацию;
- проверить на наличие пропущенных значений и дубликатов;
- провести исследовательский анализ данных;
- сделать выводы и рекомендации.

**Исходные данные.**

- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита

## Шаг. Загрузка данных¶

In [1]:
# импортируем библиотеки Pyhton для проведения исследования
import pandas as pd

### Считывание CSV-файлов с данными с помощью библиотеки pandas

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [2]:
# считали файл из csv-файла games.csv в переменную data
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

### Вывод и изучение общей информации о датафрейме

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
# вывели 20 строк датафрейма
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
# вывели информацию по датафрейму
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В рассматриваемом датафрейме 12 столбцов и 21525 строк. Есть пропущенные значения в столбцах days_employed и total_income. Необходимо заменить вещественный тип данных в столбце total_income на целочисленный. Также проверим на дубликаты и аномальные значения.

## Шаг. Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
# вывели количество пропущенных значений в столбцах
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения в данном исследовании сложно, поэтому оставим их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
# изменили тип данных в столбце
data['total_income'] = data['total_income'].astype(int)

In [15]:
# проверили изменение типа
data.dtypes

children              int64
days_employed       float64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income          int32
purpose              object
dtype: object

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [16]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [17]:
#вывели количество дубликатов
data.duplicated().sum()

71

In [18]:
#удалили дуюликаты
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [19]:
#написали функцию, которая назначает категорию в зависимости от дохода
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [20]:
#создали столбец total_income_category с категориями применив функцию
data['total_income_category'] = data['total_income'].apply(categorize_income)
data.head() # вывели 5 строк датафрейма

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category  
0               покупка жилья                     B  
1     приобретение автомобиля                     C  
2               покупка жилья                     C  
3  дополнительное образование                     B  
4             сыграть свадьбу                     C

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [21]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [22]:
#написали функцию, которая назначает категорию в зависимости от цели кредита
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [23]:
#создали столбец purpose_category с категориями применив функцию
data['purpose_category'] = data['purpose'].apply(categorize_purpose)
data.head() # вывели 5 строк датафрейма

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

На данном этапе предобработали данные: обработали пропуски в столбцах days_employed и total_income, изменили тип данных в столбце total_income, удалили дубликаты, обработали аномальные значения в столбце children. Категоризировали данные в зависимости от дохода и целей получения кредита.

## Исследуйте данные и ответьте на вопросы

В данном разделе ответим на поставленные вопросы.

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [24]:
data_debt = data['debt'].sum() / data['debt'].count()  
print('Средний процент просрочки по датасету {:.1%}'.format(data_debt)) 
# рассчитали долю случаев задолженности по возврату кредитов во всех выдачах кредитов, будем считать как среднию величину.

Средний процент просрочки по датасету 8.1%


In [25]:
'''создадим фунцию, которая будет группировать по необходимому столбцу, выводить сводную таблицу: 
количество всех кредитов, количество случаев задолженности по возврату кредитов,
долю случаев задолженности по возврату кредитов'''
def share_debt(x):
    return '{:.2%}'.format(x.sum() / x.count())
def data_pivot(column):
    data_column = data.pivot_table(index=column, values='debt', aggfunc=['count', 'sum', share_debt])
    return data_column.sort_values([('share_debt', 'debt')], ascending=False)

In [26]:
data_pivot('children')
# показали выдачу кредитов, случаи задолженности по возврату кредитов, долю просрочки в зависимости от количества детей

count   sum share_debt
           debt  debt       debt
children                        
4            41     4      9.76%
2          2052   194      9.45%
1          4808   444      9.23%
3           330    27      8.18%
0         14091  1063      7.54%
5             9     0      0.00%

In [27]:
# расчитали влияение детей на случаи задолженности по возврату кредитов
# от общей доли всех случаев задолженности по возврату кредитов отняли долю случаев в зависимости от количества детей
for d in data['children'].sort_values().unique():
    share_debt_children = data.loc[data['children'] == d, 'debt'].sum() / data.loc[data['children'] == d, 'debt'].count()
    if share_debt_children > data_debt:
        print('Количество детей: {} влияет на увеличение случаев задолженности по возврату кредитов на разницу {:.1%} от средней величины'.format(d, abs(share_debt_children-data_debt)))
    elif share_debt_children < data_debt:
        print('Количество детей: {} влияет на уменьшение случаев задолженности по возврату кредитов на разницу {:.1%} от средней величины'.format(d, abs(share_debt_children-data_debt)))
    else:
        print('Количество детей не влияет на случаи задолженности по возврату кредитов')

Количество детей: 0 влияет на уменьшение случаев задолженности по возврату кредитов на разницу 0.6% от средней величины
Количество детей: 1 влияет на увеличение случаев задолженности по возврату кредитов на разницу 1.1% от средней величины
Количество детей: 2 влияет на увеличение случаев задолженности по возврату кредитов на разницу 1.3% от средней величины
Количество детей: 3 влияет на увеличение случаев задолженности по возврату кредитов на разницу 0.1% от средней величины
Количество детей: 4 влияет на увеличение случаев задолженности по возврату кредитов на разницу 1.6% от средней величины
Количество детей: 5 влияет на уменьшение случаев задолженности по возврату кредитов на разницу 8.1% от средней величины


In [28]:
data.loc[data['children'] == 5] # вывели клиентов, у которых 5 детей и отсутсвуют случаи задолженности по возврату кредитов.

children  days_employed  dob_years  education  education_id  \
3979          5    1572.328285         42    среднее             1   
4397          5    3248.839837         36    среднее             1   
7866          5     773.124856         36    среднее             1   
15822         5     418.199982         31    среднее             1   
15916         5    2286.262752         37    среднее             1   
16211         5     387.317579         35    среднее             1   
20452         5     268.425464         38  начальное             3   
20837         5    2386.600221         35    среднее             1   
21156         5    1690.018117         59    среднее             1   

          family_status  family_status_id gender  income_type  debt  \
3979   гражданский брак                 1      M    сотрудник     0   
4397    женат / замужем                 0      F    компаньон     0   
7866    женат / замужем                 0      F    сотрудник     0   
15822   женат / замужем                 0      F    сотрудник     0   
15916   женат / замужем                 0      F    сотрудник     0   
16211  гражданский брак                 1      F  госслужащий     0   
20452   женат / замужем                 0      F    сотрудник     0   
20837   женат / замужем                 0      F    компаньон     0   
21156   женат / замужем                 0      M    сотрудник     0   

       total_income                           purpose total_income_category  \
3979         142594      на покупку своего автомобиля                     C   
4397         168460          операции с недвижимостью                     C   
7866          48772                 операции с жильем                     D   
15822         77552  сделка с подержанным автомобилем                     C   
15916        256698              покупка недвижимости                     B   
16211        126102             на проведение свадьбы                     C   
20452        212545      заняться высшим образованием                     B   
20837        204241                             жилье                     B   
21156        269068   операции со своей недвижимостью                     B   

               purpose_category  
3979     операции с автомобилем  
4397   операции с недвижимостью  
7866   операции с недвижимостью  
15822    операции с автомобилем  
15916  операции с недвижимостью  
16211        проведение свадьбы  
20452     получение образования  
20837  операции с недвижимостью  
21156  операции с недвижимостью

Исходя из расчитанных данных можно выделить следующее: существует положительная зависимость между количеством детей и возвратом кредита в срок. У клиентов у которых есть дети, случаи не возврата кредита в срок выше по сравнению со средней величиной, диапазон разницы колеблется от 0.1% до 1.6%. Аномалия у клиентов у которых 5 детей - отсутствуют случаи задолженности по возврату кредитов, таких клиентов из всего датасета 9. У таких клиентов можно выделить уровень дохода средний и выше среднего, что могло сказаться на платежеспособности. Также хочется отметить снижение количества кредитов в зависимости от количества детей от 14091 заявок клиентов, у которых нет детей и всего 9 заявок у клиентов с 5 детьми.
Сравнив сопоставимые по размеру катогории заемщиков с 1 и 2 детьми и заемщиков без детей можно выявить, что доля просрочки у заемщиков без детей ниже, чем у заемщиков с 1 или 2 детьми. Разница составляет 1,8% между данными категориями.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [29]:
data_pivot('family_status')
# показали выдачу кредитов, случаи задолженности по возврату кредитов, долю просрочки в зависимости от семейного статуса

count  sum share_debt
                        debt debt       debt
family_status                               
Не женат / не замужем   2796  273      9.76%
гражданский брак        4134  385      9.31%
женат / замужем        12261  927      7.56%
в разводе               1189   84      7.06%
вдовец / вдова           951   63      6.62%

In [30]:
# расчитали влияение семейного статуса на случаи задолженности по возврату кредитов
# от общей доли всех случаев задолженности по возврату кредитов отняли долю случаев в зависимости от семейного статуса
for d in data['family_status'].sort_values().unique():
    share_debt_family = data.loc[data['family_status'] == d, 'debt'].sum() / data.loc[data['family_status'] == d, 'debt'].count()
    if share_debt_family > data_debt:
        print('Семейный статус: {} влияет на увеличение случаев задолженности по возврату кредитов на разницу {:.1%} от средней величины'.format(d, abs(share_debt_family-data_debt)))
    elif share_debt_family < data_debt:
        print('Семейный статус: {} влияет на уменьшение случаев задолженности по возврату кредитов на разницу {:.1%} от средней величины'.format(d, abs(share_debt_family-data_debt)))
    else:
        print('Семейный статус не влияет на случаи задолженности по возврату кредитов') 

Семейный статус: Не женат / не замужем влияет на увеличение случаев задолженности по возврату кредитов на разницу 1.6% от средней величины
Семейный статус: в разводе влияет на уменьшение случаев задолженности по возврату кредитов на разницу 1.1% от средней величины
Семейный статус: вдовец / вдова влияет на уменьшение случаев задолженности по возврату кредитов на разницу 1.5% от средней величины
Семейный статус: гражданский брак влияет на увеличение случаев задолженности по возврату кредитов на разницу 1.2% от средней величины
Семейный статус: женат / замужем влияет на уменьшение случаев задолженности по возврату кредитов на разницу 0.6% от средней величины


Иcходя из расчитанных данных можно выявить зависимость между семейным положением и возвратом кредита в срок. У клиентов с семейным статусом: женат / замужем, в разводе, вдовец / вдова меньше случаев не возвращения кредита в срок. Разница от средней величины колеблеться от 0,6% до 1,5%. У клиентов с семейным стутасом Не женат / не замужем либо гражданский брак наблюдается увеличение случаев задолженности по возврату кредитов в срок на 1.6% и 1.2% соответственно от средней величины случаев. Обобщив категории по отношению к официальному браку можно выявить, что клиенты с узаконенными отношениями меньше допускают просрочек, разница составляет почти 2%, если объеденить категории не узаконенных отношений в гражданском браке и категорию не женат / не замужем.   

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [31]:
data_pivot('total_income_category')
# показали выдачу кредитов, случаи задолженности по возврату кредитов, долю просрочки в зависимости от уровня дохода

count   sum share_debt
                        debt  debt       debt
total_income_category                        
E                         22     2      9.09%
C                      15921  1353      8.50%
A                         25     2      8.00%
B                       5014   354      7.06%
D                        349    21      6.02%

In [32]:
# расчитали влияение уровня дохода на случаи задолженности по возврату кредитов
# от общей доли всех случаев задолженности по возврату кредитов отняли долю случаев в зависимости от уровня дохода
for d in data['total_income_category'].sort_values().unique():
    share_debt_income = data.loc[data['total_income_category'] == d, 'debt'].sum() / data.loc[data['total_income_category'] == d, 'debt'].count()
    if share_debt_income > data_debt:
        print('Уровень дохода: {} влияет на увеличение случаев задолженности по возврату кредитов на разницу {:.1%} от средней величины'.format(d, abs(share_debt_income-data_debt)))
    elif share_debt_income < data_debt:
        print('Уровень дохода: {} влияет на уменьшение случаев задолженности по возврату кредитов на разницу {:.1%} от средней величины'.format(d, abs(share_debt_income-data_debt)))
    else:
        print('Уровень дохода не влияет на случаи задолженности по возврату кредитов')

Уровень дохода: A влияет на уменьшение случаев задолженности по возврату кредитов на разницу 0.1% от средней величины
Уровень дохода: B влияет на уменьшение случаев задолженности по возврату кредитов на разницу 1.1% от средней величины
Уровень дохода: C влияет на увеличение случаев задолженности по возврату кредитов на разницу 0.4% от средней величины
Уровень дохода: D влияет на уменьшение случаев задолженности по возврату кредитов на разницу 2.1% от средней величины
Уровень дохода: E влияет на увеличение случаев задолженности по возврату кредитов на разницу 1.0% от средней величины


Иcходя из расчитанных данных можно выявить зависимость между уровнем дохода и возвратом кредита в срок. Самое большое отклонение у клиентов с низким уровнем дохода 0–30000 - отклонение от средней величины на 1%. У остальных категорий значения около среднего, либо ниже. Самое большое количество кредитов у клиентов с уровнем дохода 50001–200000 - количество заявок составляет 15921, что более чем в 3 раза больше следующей категории. Сравнив уровень дохода только двух самых многочисленных выборок С и В можно выявить, что чем выше уровень дохода, тем ниже уровень просрочки: разница между категориями 1,43%.

### Как разные цели кредита влияют на его возврат в срок?

In [33]:
data_pivot('purpose_category')
# показали выдачу кредитов, случаи задолженности по возврату кредитов, долю просрочки в зависимости от целей кредита

count  sum share_debt
                           debt debt       debt
purpose_category                               
операции с автомобилем     4279  400      9.35%
получение образования      3988  369      9.25%
проведение свадьбы         2313  183      7.91%
операции с недвижимостью  10751  780      7.26%

In [34]:
# расчитали влияение целей кредита на случаи задолженности по возврату кредитов
# от общей доли всех случаев задолженности по возврату кредитов отняли долю случаев в зависимости от уровня дохода
for d in data['purpose_category'].sort_values().unique():
    share_debt_purpose = data.loc[data['purpose_category'] == d, 'debt'].sum() / data.loc[data['purpose_category'] == d, 'debt'].count()
    if share_debt_purpose > data_debt:
        print('Цель кредита: {} влияет на увеличение случаев задолженности по возврату кредитов на разницу {:.1%} от средней величины'.format(d, abs(share_debt_purpose-data_debt)))
    elif share_debt_purpose < data_debt:
        print('Цель кредита: {} влияет на уменьшение случаев задолженности по возврату кредитов на разницу {:.1%} от средней величины'.format(d, abs(share_debt_purpose-data_debt)))
    else:
        print('Цель кредита не влияет на случаи задолженности по возврату кредитов')

Цель кредита: операции с автомобилем влияет на увеличение случаев задолженности по возврату кредитов на разницу 1.2% от средней величины
Цель кредита: операции с недвижимостью влияет на уменьшение случаев задолженности по возврату кредитов на разницу 0.9% от средней величины
Цель кредита: получение образования влияет на увеличение случаев задолженности по возврату кредитов на разницу 1.1% от средней величины
Цель кредита: проведение свадьбы влияет на уменьшение случаев задолженности по возврату кредитов на разницу 0.2% от средней величины


Иcходя из расчитанных данных можно выявить зависимость между целями кредита и возвратом кредита в срок. По целям операции с автомобилем и получение образования выявлено увеличение случаев задолженности по возврату кредитов в срок, отклонение от средней величины на 1,2% и 1,1% соответственно. По целям проведение свадьбы и операции с недвижимостью выявлено снижение случаев задолженности по возврату кредитов в срок, отклонение от средней величины на 0,2% и 0,9% соответственно. Несмотря на то, что у цели операции с недвижимостью наибольшее количество заявок 10751, доля просрочки в этой категории ниже чем у остальных.

***У клиентов с уровнем дохода до 30 тыс.руб, а также с доходом от 50-200 тыс.руб. с семейным статусом не женат / не замужем либо гражданский брак при наличии детей, которые берут денежные средства в кредит на операции с автомобилем либо получение образования доля просрочки выше средней по представленным данным.  
Сделав фильтр по данным категориям можно получить долю почти 16% (расчет ниже), что практически в 2 раза выше средней по датасету 8,1%***

In [35]:
data.loc[(data['children'] > 0) & (data['children'] != 5) \
& ((data['total_income_category'] == 'C') | (data['total_income_category'] == 'E')) \
& ((data['family_status'] == 'Не женат / не замужем') | (data['family_status'] == 'гражданский брак')) \
& ((data['purpose_category'] == 'операции с автомобилем') | (data['purpose_category'] == 'получение образования'))]['debt'].sum() \
/ data.loc[(data['children'] > 0) & (data['children'] != 5) \
& ((data['total_income_category'] == 'C') | (data['total_income_category'] == 'E')) \
& ((data['family_status'] == 'Не женат / не замужем') | (data['family_status'] == 'гражданский брак')) \
& ((data['purpose_category'] == 'операции с автомобилем') | (data['purpose_category'] == 'получение образования'))]['debt'].count()*100

15.96009975062344

***У клиентов с доходом от 30 тыс.руб. за исключением категории дохода от 50-200 тыс.руб. с семейным статусом женат/замужем, в разводе либо вдовец / вдова с 5 детьми или без детей, которые берут денежные средства в кредит на проведение свадьбы 
либо операции с недвижимостью доля просрочки ниже, чем средняя по представленным данным.   
Настроив фильтр по данным категориям получим долю просрочки в 7,5%, на 0,6% ниже средней по датасету.***

In [36]:
data.loc[(data['children'] == 0) | (data['children'] == 5) \
& ((data['total_income_category'] == 'A') | (data['total_income_category'] == 'B') | (data['total_income_category'] == 'D')) \
& ((data['family_status'] == 'женат / замужем') | (data['family_status'] == 'в разводе') | (data['family_status'] == 'вдовец / вдова')) \
& ((data['purpose_category'] == 'проведение свадьбы') | (data['purpose_category'] == 'операции с недвижимостью'))]['debt'].sum() \
/ data.loc[(data['children'] == 0) | (data['children'] == 5) \
& ((data['total_income_category'] == 'A') | (data['total_income_category'] == 'B') | (data['total_income_category'] == 'D')) \
& ((data['family_status'] == 'женат / замужем') | (data['family_status'] == 'в разводе') | (data['family_status'] == 'вдовец / вдова')) \
& ((data['purpose_category'] == 'проведение свадьбы') | (data['purpose_category'] == 'операции с недвижимостью'))]['debt'].count()*100

7.541681447321745

### Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Причины могут быть технологические - при выгрузке какие-то данные были утеряны, либы человеческие - сам клиент не заполнил данные поля, а система пропустила заявку далее, либо этих данных просто нет, т.к. предложения по кредитам могут формироваться автоматически и данные поля не обязательны. Можно проверить с разработчиками выгрузку данных, а также настроить систему, чтобы при не заполнении поля, подсвечивало данное поле, просило заполнить и не пропускало далее. В случае подачи заявки самостоятельно через сайт либо отклик на предложение, этот момент лучше проверить на ограниченном числе клиентов, есть вероятность снижения количества заявок, т.к. клиент может бросать заполнение заявки. 

### Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* В данных могут быть выбросы - либо очень маленькое значение, либо очень большое, что значительно скажется на средней величине и введет в заблужение о состоянии количественных переменных.

## Вывод

Для проведения иследования была получена статистика о платёжеспособности клиентов, где по заемщикам было указана гендерная принадлежность, семейное положение и количество детей, уровень дохода и образования, возраст клиентов, тип занятости и их общий трудовой стаж, цели кредитования и факт того, имел ли задолженность по возврату кредитов.

На стадии предобработки данных были выявлены пропуски в столбцах с общим трудовым стажем и уровнем дохода, также были выявлены дубликаты в данных и аномально большие значения в столбцах с общим трудовым стажем и количеством детей.

Дубликаты и аномально большие значения в столбце с количеством детей были удалены. У двух типов (безработные и пенсионеры) аномально большие значения в столбце с общим трудовым стажем исправить сложно, поэтому оставили их как есть. Пропущенные значения были заполнены медианым значением с разбивкой на категории по типу занятости.

Для проверки гипотезы построили сводные таблицы по запрошенным категориям, где необходимо было выявить зависимости. В сводные таблицы добавили столбцы с долей просрочки по каждой из категорий.

В результате исследования были выявлены зависимости во всех запрошенных категориях на уровень просрочки. Заемщики без детей допускали меньше просрочек по кредиту, чем клиенты с детьми. Заемщики в официальном браке и узаконенных отношениях также обладают меньшей долей случаев задолженности по кредиту, чем клиенты в гражданском браке или со статусом не женат / не замужем. Доля просрочки снижается с ростом уровня дохода заемщиков. Также оказали влияние и цели кредитования: у сделок операции с недвижимостью и проведение свадьбы уровень случаев задолженности по кредиту ниже, чем у сделок операции с автомобилем и получение образования.

Таким образом гипотеза исследования полностью подтвердилась.